### DISCUSSION TITLE
Patch 3.13.6 broke x-effect

### DISCUSSION THREAD

**Discussion by shanekunz** [alpinejs/alpine#4080] [7. Bugs (archived)] [answered]
I've been developing using 3.x.x (`<script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.x.x/dist/cdn.min.js"></script>`)
Today I woke up and my code was making infinite network calls.
I determined that in the jump from 3.13.5 -> 3.13.6, `x-effect` is no longer intelligently determining when to run.

For example I have:
`x-effect="getDailyAttendanceRecords(year, month, clientId, programId, providerId)"`
...
Inside my `Alpine.data('attendanceData', () => ({ ` see the "BUG" comment
```
getDailyAttendanceRecords(year, month, clientId, programId, providerId) {
  const alpineData = this;
  alpineData.loading = true;
  fetch(`${baseUrl}?handler=GetDailyAttendanceRecords&year=${year}&month=${month}&clientId=${clientId}&programId=${programId}&providerId=${providerId}`)
      .then(response => response.json())
      .then(data => {
        alpineData.loading = false; // BUG! This causes the function to rerun causing an infinite loop despite "loading" not being referenced in the x-effect attribute.
        console.log('infinite loop?');
        return;
      })
}
```
I tried using `this` directly and it also didn't work. It seems to have something to do with it being inside of a promise for whatever reason, but I don't see a reason that should affect `x-effect` logic. 

Going down one patch number to 3.13.5 fixes the issue for me entirely.

Let me know if you want more code, I can send privately.

**Comment by ekwoka** [comment]
You do reference loading in the x-effect.

```
alpineData.loading = true
```

🤔

It shouldn't trigger tracking...

And I'm not seeing anything that should be causing this.... :/


Might be related to the changes to merge proxies that is being eager in registering dependents...but it doesn't seem like it should...

**Reply by shanekunz** [reply]
Yeah, sorry, I might struggle with the terminology a little, I updated my comment to say x-effect _attribute_. How I understand it from react.js is that whatever is referenced in the x-effect attribute is essentially the `dependencies` in `useEffect(setup, dependencies?)` and anything inside the `setup` function doesn't trigger itself unless of course it changes something in `dependencies`.

Maybe I'm trying to do something too advanced here, as the docs aren't clear about what should happen in this scenario. However, the fact that the functionality completely flipped after a patch release is the concerning part. If I have time tomorrow I could try to translate this into a `$watch` magic method, however, I think that would also require me to refactor the variables into a single object.

Worst case, I'm comfortable just leaving it at `3.13.5`. Thankfully this project isn't live yet, and I shouldn't reference a wildcard cdn file on a production site anyways. Let me know if there's anything I can do to help, I don't work on open source, but I'm happy to give it a shot, as I'd prefer to use Alpine as the main JS tool with our .NET sites.

**Reply by ekwoka** [reply]
React is a bad mental model to look at.

the "dependencies" are any value read/accessed while executing the synchronous portion of the expression.

**Reply by shanekunz** [reply]
Ok I see what you're saying, the assignment outside the `fetch` starts the watch on `loading`. I validated that running code.

However, your synchronous explanation is still not entirely true with the latest version. I wrapped the entire function contents with an async method:
```
getDailyAttendanceRecords(year, month, clientId, programId, providerId) {
    (async () => {
       ...
    })();
}
```
And it still watches the contents inside the `async` function but outside the `fetch` and creates an infinite loop. Timing does seem to be relevant here though because hitting break points fixes it, which probably simulates the delay of a network request.

I will see if `$watch` works now if I put `year, month, clientId, programId, providerId` into a single `searchParams` alpine data object and watch it.

**Reply by SimoTod** [reply]
Async without a wait doesn't make the function run outside of the effect. Use Alpine.nextTick

**Reply by ekwoka** [reply]
Wrapping a function in "async" doesn't make the execution of it asynchronous. It it makes the RESOLUTION of it asynchronous.

Promises are executed eagerly in javascript.

So when you call an async function, it is executed exactly like a sync function until it hits an await inside of itself. and the return is wrapped in a Promise.

But yet, I made a PR to fix it. Due to a change in how the proxy stack was handled, this caused sets to hook into reacticity, which they shouldn't have, but was unclear that there was a difference.

#4081 will resolve this.

Also adds a test for it.

It was a lapse in checking, since some of this internal reactivity stuff is tricky, especially when Alpine doesn't use it's own reactivity engine to more easily track the innards of.

**Reply by shanekunz** [reply]
@ekwoka and @SimoTod thanks y'all. `$watch` ended up doing the trick on the latest Alpine version. It's good to know about how `async` works in JavaScript, but I'm going to stick with the `$watch` solution since it's more explicit. Ultimately, I probably shouldn't have used `x-effect` for this in the first place, but hopefully finding the behavior difference between patch versions and @ekwoka 's new patch will help improve the reactivity system going forward.

```
document.addEventListener('alpine:init', () => {
    Alpine.data('attendanceData', () => ({
...
        search: {
            year: initYear,
            month: initMonth,
            clientId: initClientId,
            programId: initProgramId,
            providerId: initProviderId
        }
...
        init() {
            const alpineData = this;
...
            alpineData.$watch('search', (value) => {
                alpineData.getDailyAttendanceRecords(value);
            });
            alpineData.getDailyAttendanceRecords(value);
        }
...
```

**Reply by ekwoka** [reply, marked as answer]
I actually favor using effect.

The main issue in this will work in 3.13.5 and 3.13.7. .6 just has a bug.

Setting should not register as a dependency.

**Reply by shanekunz** [reply]
Oh I see, 'Setting' meaning like `alpineDataValue = true`. I'll use `x-effect` next time then. Thanks for the prompt fix!

**Reply by ekwoka** [reply]
Yes, setting a value to the context.
